src/shogun/io/SGIO.h

```diff
@@ -328,89 +300,45 @@ class SGIO
 		void done();
 
 		/** print error message 'not implemented' */
-		inline void not_implemented(const char* function, const char* file, int32_t line) const
+		inline void not_implemented(const SourceLocation& loc={}) const
 		{
 			error<ShogunException>(
-			    MSG_ERROR, function, file, line,
-			    "Sorry, not yet implemented .\n");
+			    MSG_ERROR, loc, "Sorry, not yet implemented .\n");
 		}
 
 		/** print error message 'Only available with GPL parts.' */
-		inline void gpl_only(const char* function, const char* file, int32_t line) const
+		inline void gpl_only(const SourceLocation& loc={}) const
 		{
 			error<ShogunException>(
-			    MSG_ERROR, function, file, line, "This feature is only "
-			                                     "available if Shogun is built "
-			                                     "with GPL codes.\n");
+			    MSG_ERROR, loc,
+			    "This feature is only "
+			    "available if Shogun is built "
+			    "with GPL codes.\n");
 		}
 
 		/** print warning message 'function deprecated' */
-		inline void deprecated(const char* function, const char* file, int32_t line) const
+		inline void deprecated(const SourceLocation& loc={}) const
 		{
-			message(MSG_WARN, function, file, line,
+			message(MSG_WARN, loc,
 					"This function is deprecated and will be removed soon.\n");
 		}
 
-		/** print a buffered message
-		 *
-		 * @param prio message priority
-		 * @param fmt format string
-		 */
-		void buffered_message(EMessageType prio, const char *fmt, ... ) const;
-
-		/** skip leading spaces
-		 *
-		 * @param str string in which to look for spaces
-		 * @return string after after skipping leading spaces
-		 */
-		static char* skip_spaces(char* str);
-
-		/** skip leading spaces + tabs
-		 *
-		 * @param str string in which to look for blanks
-		 * @return string after after skipping leading blanks
-		 */
-		static char* skip_blanks(char* str);
-
-		/** get target
-		 *
-		 * @return file descriptor for target
-		 */
-		inline FILE* get_target() const
-		{
-			return target;
-		}
-
-		/** set target
-		 *
-		 * @param target file descriptor for target
-		 */
-		void set_target(FILE* target);
-
-		/** set target to stderr */
-		inline void set_target_to_stderr() { set_target(stderr); }
+		/** redirects stdout to another sink */
+		void redirect_stdout(std::shared_ptr<spdlog::sinks::sink> sink);
 
-		/** set target to stdout */
-		inline void set_target_to_stdout() { set_target(stdout); }
+		/** redirects stderr to another sink */
+		void redirect_stderr(std::shared_ptr<spdlog::sinks::sink> sink);
```

Comment:
there are only two sinks: one for stdout and the other for stderr. if the user wants to redirect to multiple sinks, he could use "dist_sink"